chalk/trail.py

```diff
@@ -1,110 +1,166 @@
+
+SegmentLike = Union[Segment, ArcSegment]
 
 
 @dataclass
-class Trail(Transformable):
-    """Trail class.
+class Located(Enveloped, Traceable, Transformable):
+    trail: Trail
+    location: P2
 
-    This is derived from a ``chalk.transform.Transformable`` class.
+    def located_segments(self) -> Iterable[Tuple[SegmentLike, P2]]:
+        return zip(self.trail.segments, self.points())
 
-    [TODO]: Need more explanation on what this class is for (preferably
-    with illustrations/figures).
-    """
+    def points(self) -> Iterable[P2]:
+        return (pt + self.location for pt in self.trail.points())
 
-    offsets: List[V2]
+    def get_envelope(self) -> Envelope:
+        return Envelope.concat(
+            segment.get_envelope().translate_by(location)
+            for segment, location in self.located_segments()
+        )
+
+    def get_trace(self) -> Trace:
+        return Trace.concat(
+            segment.get_trace().translate_by(location)
+            for segment, location in self.located_segments()
+        )
+
+    def stroke(self) -> Diagram:
+        return self.to_path().stroke()
+
+    def apply_transform(self, t: Affine) -> Located:  # type: ignore
+        t2 = remove_linear(t)
+        return Located(
+            apply_affine(t, self.trail), apply_affine(t2, self.location)
+        )
 
+    def to_path(self) -> Path:
+        from chalk.shapes.path import Path
+
+        return Path([self])
+
+
+@dataclass
+class Trail(Transformable, TrailLike):
+    segments: List[SegmentLike]
+    closed: bool = False
+
+    # Monoid - concat
     @staticmethod
     def empty() -> Trail:
-        return Trail([])
+        return Trail([], False)
 
     def __add__(self, other: Trail) -> Trail:
-        """Adds another trail to this one and
-        returns the resulting trail.
-
-        Args:
-            other (Trail): Another trail object.
-
-        Returns:
-            Trail: A trail object.
-        """
-        new_vec = list(self.offsets)
-        new_vec.extend(other.offsets)
-        return Trail(new_vec)
-
-    @classmethod
-    def from_path(cls, path: Path) -> Trail:
-        """Constructs and returns a trail from a given path.
-
-        Args:
-            path (Path): A path object.
-
-        Returns:
-            Trail: A trail object.
-        """
-        pts = path.points()
-        offsets = [t - s for s, t in zip(pts, pts[1:])]
-        return cls(offsets)
-
-    def to_path(self, origin: P2 = P2(0, 0)) -> Path:
-        """Converts a trail to a path, given a point (as a reference).
-
-        Args:
-            origin (P2, optional): A point object.
-                                      Defaults to ORIGIN.
-
-        Returns:
-            Path: A path object.
-        """
-        points = [origin]
-        for s in self.offsets:
-            points.append(points[-1] + s)
-        return Path.from_points(points)
+        return Trail(self.segments + other.segments, False)
```

Comment:
The combination of two trails when one of them is a loop (that is, closed) might not be well defined. I see that in Haskell they have explicitly disallowed the `Monoid` instance for `Trail' Loop v n` and I guess that's one of the reasons why they make the explicit distinction of the two types of `Trail`s (lines and loops) in the type system:
```haskell
-- | The empty trail is constantly the zero vector.  Trails are
--   composed via concatenation.  Note that only lines have a monoid
--   instance (and not loops).
instance (Metric v, OrderedField n) => Monoid (Trail' Line v n) where
  mempty  = emptyLine
  mappend = (<>)
```
I think I'm fine with our approach, but we should at least make a special note in the documentation such that we don't forget about this design choice.

Reply:
Oh that's really interesting. I was wondering why they made that type distinction, but this makes a lot of sense. 

In python, I'll just fail if you try to mappend a loop?

Reply:
> In python, I'll just fail if you try to mappend a loop?

Yes, I guess raising an exception when `mappend`ing a loop would be the safest way of avoiding unexpected behavior.